### ISSUE TITLE
CompoundAssignments: check for compatible types in assignment 

### ISSUE THREAD

**Issue by smillst** [typetools/checker-framework#624]
Given the following code:
````
String s += expr;
```
This assignment is legal if the type of `s + expr` is a subtype of the type of `s`.  However, BaseTypeVisitor actually checks that the type of `expr` is a subtype of `s`. 

In BaseTypeVisitor:
````
 @Override
    public Void visitCompoundAssignment(CompoundAssignmentTree node, Void p) {
        commonAssignmentCheck(node.getVariable(), node.getExpression(),
                "compound.assignment.type.incompatible");
        return super.visitCompoundAssignment(node, p);
    }
````
Both the Signature Checker (in PR #612) and the Lock Checker (in PR #585) have work arounds for this.  And the Value Checker should issue a warning for the code below, but doesn't because of this bug:

        @StringVal("hello") String s = "hello";
        s += "hello";

**Comment by smillst** [comment]
This compound assignment in the Format Checker should also issue an error:

        @Format({ ConversionCategory.GENERAL, ConversionCategory.INT}) 
        String f = "String '%s' has length %d";
        f +=  "String '%s' has length %d";

**Comment by wmdietl** [comment]
Is the type that we determine for the compound assignment correct, as found in:
https://github.com/typetools/checker-framework/blob/master/framework/src/org/checkerframework/framework/type/treeannotator/PropagationTreeAnnotator.java#L112

it seems to compute the LUB of `s` and `expr`.
If so, would the following fix the issue:

```
    @Override
    public Void visitCompoundAssignment(CompoundAssignmentTree node, Void p) {
        commonAssignmentCheck(node.getVariable(), node,
                "compound.assignment.type.incompatible");
        return super.visitCompoundAssignment(node, p);
    }
```

That is, test the type of the overall tree against the type of the LHS.

Then a type system developer would only need to adapt the type in the ATF to change the behavior.

**Comment by smillst** [comment]
I've tried to come up with a scenario where the type of `s += expr` is not the same as the type of `s + expr`, but I can't think of one. So, I think this will work.   

I've tested this on the aft-clean-up branch and after a fix to the Value Checker, the correct error is issued. 

The Format Checker does not calculate the type of a string concatenation correctly, so that will have to be fixed before it can be tested.

**Comment by mernst** [comment]
> I've tried to come up with a scenario where the type of s += expr is not the same as the type of s + expr, but I can't think of one. So, I think this will work.

I agree.
